### PR TITLE
KEP - Establish KEP process

### PR DESCRIPTION
**Pull request by nrfox**
Establishes a process to propose enhancements to the Kiali project.

[discussion](https://github.com/kiali/kiali/discussions/4208)

### CODE REVIEW

File: design/KEPS/kiali-enhancement-proposal/proposal.md
Comment (lucasponce) [inline design/KEPS/kiali-enhancement-proposal/proposal.md:49]: > Proposals should strive to reach consensus by the end of the next sprint.

I have feeling that set a deadline to have a consensus may introduce pressure, I'd propose to reword this to have "progress".

What do I mean ? I'd like if a proposal is quite active even it hasn't reached any consensus.
I.e.:
- There is a first design 1 proposed by folk A
- Folk B, reviews and proposes a design 2, reviewed by Folk A.
- Iteration over the previous step, even if there is no consensus, but there is progress.

It may involve more than a Sprint.

In general, a proposal shouldn't take too long, but I have the impression that there are areas that may need time if we want to have a good proposal, discussion and consensus.

Examples:
- Redesign of APIs (introducing GraphQL instead of pure REST).
- Redesign of new backend model.
- Discussion about what is the scope of a service (health) and how these services depend on others internal components.
- Push vs pull model in the UI.

Just to cite some areas that are suitable for KEP but I they deserves a detailed discussion.

Comment (lucasponce) [reply design/KEPS/kiali-enhancement-proposal/proposal.md:49]: Personally I guess that time boxing is something that add pressure and makes that if the last day of the deadline we can rush. 
I think that the commitment should be that maintainers should get the commitment to keep active KEP until that is resolved.
To me the "active" verb is more flexible (it should be stalled) and also doesn't box something to rush that in a sprint there is a decision.
Also, in case that there are any need to poll, the governance is still there.
I'd try to avoid to put more time boxing rather than the formal one defined in the governance doc.

I think that when there is a relevant KEP proposed is in the maintainers interestest to participate, and also in the owners to seek consensus.

Comment (nrfox) [reply design/KEPS/kiali-enhancement-proposal/proposal.md:49]: I'd say the goal is less to add pressure and more to set the expectation that the proposals should have active involvement and eventually resolve instead of staying open forever. I'm ok with removing the timeboxing but perhaps we can also put a cap on the number of active proposals there are at any given time like what @jshaughn [suggested](https://github.com/kiali/kiali/discussions/4208#discussioncomment-1084133)?

Comment (lucasponce) [reply design/KEPS/kiali-enhancement-proposal/proposal.md:49]: +1 saying that the number of active proposals can be limited and others may be queue is a right topic, to answer a question "why my proposal is not yet started", "because, mantainers are reviewing A,B,C"

Comment (nrfox) [reply design/KEPS/kiali-enhancement-proposal/proposal.md]: Alright I didn't propose a hard limit on the number of active proposals but merely stated that priority would be given to existing proposals over new ones.

**Review by lucasponce — APPROVED** [review]
Thanks @nrfox for the discussion.
I see all my comments addressed and the process looks good to me.

**Comment by lucasponce** [comment]
Has any other @kiali/maintainers any comment about this proposal ?

**Review by leandroberetta — APPROVED** [review]
LGTM, thanks @nrfox for this document.

**Review by jshaughn — APPROVED** [review]
@nrfox This looks good to me, I'll approve.  One question I have is about planning work/tracking. 
 This goes straight from discussion to KEP PR but a KEP could be a significant effort.  Is there any suggestion for how to work this into sprint planning?  Perhaps an empty Draft PR could be generated, assigned and added to the sprint?

**Comment by lucasponce** [comment]
> This goes straight from discussion to KEP PR but a KEP could be a significant effort. Is there any suggestion for how to work this into sprint planning? Perhaps an empty Draft PR could be generated, assigned and added to the sprint?

Reading the doc https://github.com/kiali/kiali/pull/4581/files#diff-7c0b3e560b231da3310e591fc7dab49df301acd0e5162f1e89f3d2d94cef3547R29, I think that the KEP should be a tool that may help us to progress, but not a mandatory step that may create more burden.

The Epic should be the driver, and the KEP should be a tool that can be proposed or requested when some design is not straightforward, or there is a need to discuss in more detail before starting any implementation.

At least is how I interpret the spirit of this process.

**Comment by lucasponce** [comment]
> Maybe what I said was confusing. It was just a minor point/suggestion saying that, when a KEP is being generated, there may not be an Epic/issue, but only a gh discussion.

I guess it would be nice to keep it simple, to me this is just a document used during a research phase. And it's good because that would help to understand and discuss further in the details. 

But a research should have an issue, and the issue is what we should track in the planning.

To me the ideal steps like an example:
- We have an open discussion about scalability, where several areas are touched.
- One area is to improve the topology under some scenarios.
- An Epic is created to capture that there will be work in that space, assigned and also the owner should work to progress in that area.
- As part of the Epic, a research may be needed, in that phase is were this KEP could be useful, i.e. we are talking about a "model", but perhaps it's not fully clear from the discussion, and a more technical design would help to put everyone in the same page.
- This is an intermediate step from "ok, I'll start coding and I'll let you know when I have a model ready".

Again, this is not mandatory, and not required for every aspect, only on those ones that we feel that it may help to understand the idea or to better split the Epic task into subtasks that can be assigned to the team.

**Comment by nrfox** [comment]
> Is there any suggestion for how to work this into sprint planning? Perhaps an empty Draft PR could be generated, assigned and added to the sprint?

My thoughts are that there would be an issue or epic created for the feature the KEP is addressing and the work would be tracked there instead of on the KEP PR itself. If there's enough coherent discussion around a topic to warrant a design document then there should definitely be enough to write an issue/epic to track the work.

> To me the ideal steps like an example:

    We have an open discussion about scalability, where several areas are touched.
    One area is to improve the topology under some scenarios.
    An Epic is created to capture that there will be work in that space, assigned and also the owner should work to progress in that area.
    As part of the Epic, a research may be needed, in that phase is were this KEP could be useful, i.e. we are talking about a "model", but perhaps it's not fully clear from the discussion, and a more technical design would help to put everyone in the same page.
    This is an intermediate step from "ok, I'll start coding and I'll let you know when I have a model ready".

This is pretty much what I was envisioning the progression to be as well. GH discussion --> issue/epic --> (optional) KEP design --> Implementation.

**Comment by lucasponce** [comment]
Merging.
This PR has been commented on several times to all @kiali/maintainers group during meetings; I expected more participation in the review, but at least three people agree that this could be a good idea.

Also, this is an optional tool that can be refined when needed. But I don't think it should keep it open.

**Pull request merged**
